laurat007/assignment3/Word.java

```diff
@@ -0,0 +1,243 @@
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.TreeSet;
+
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+/**
+ *
+ * @author Laura
+ * 
+ *  This program returns the words that can be formed using the letters in a 
+ * given grid by moving left, right, up, down and on diagonals. 
+ *  I used a dictionary class in order to define the correspondence between a 
+ * particular word and its explanation ( + the 2 helping methods isPrefix and 
+ * isWord, and a grid class in order to define a cell (a pair of indexes - which 
+ * facilitates moving in the grid) and its value.
+ *  The Cell subclass has methods implemented for every moving option.
+ *  Moving in the grid is similar to a backtracking from each Cell which stops
+ * if the current word formed is not a prefix, and if it finds a word it is added
+ * to solution.
+ * 
+ */
+
+class Dictionary {
+    HashMap<String, String> dictionary;
+    
+    Dictionary(HashMap<String, String> wordsSet) {
+        this.dictionary = new HashMap<>();
+        dictionary.putAll(wordsSet);
+    }
+    
+    boolean isPrefix(String prefix) {
+        Set<String> keys = dictionary.keySet();
+        Iterator it = keys.iterator();
+        while(it.hasNext()) {
+            String next = (String) it.next();
+            if(next.startsWith(prefix))
+                return true;
+        }
+        return false;
+    }
+    
+    boolean isWord(String word) {
+        Set<String> keys = dictionary.keySet();
+        return keys.contains(word);
+    }
+}
+
+class Grid {
+    
+    Dictionary dictionary;
+    int rows, columns;
+    char[][] grid;
+    Cell[][] cells;
+    
+    Grid(int r, int c, char[][] letterMatrix, Dictionary dict) {
+        dictionary = dict;
+        rows = r;
+        columns = c;
+        grid = new char[rows][columns];
+        cells = new Cell[rows][columns];
+        for(int i = 0; i < rows; i++)
+            for(int j = 0; j < columns; j++) {
+                cells[i][j] = new Cell(i, j);
+                grid[i][j] = letterMatrix[i][j];
+            }
+    }
+   
+    class Cell {
```

Comment:
I'm not sure I understand the idea behind this class at all, why can't we just move around in a grid of letters? Also the implementation of moving is very inefficient: every time you call "up()" or any other moving method you create a new instance of Cell. Can you count how many of them are created for a small grid of 2x3? It will not work well for any larger number

Reply:
I got a little confused with that Cell too. After your feedback I realised it was not needed so thank you! Replaced those moving methods with x and y indexes in grid.